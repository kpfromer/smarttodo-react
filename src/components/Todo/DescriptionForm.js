import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class DescriptionForm extends Component {

  static propTypes = {
    onUpdate: PropTypes.func.isRequired,
    description: PropTypes.string,
    className: PropTypes.string,
    invalidClassName: PropTypes.string,
    clearOnUpdate: PropTypes.bool
  };

  static defaultProps = {
    description: '',
    className: '',
    invalidClassName: '',
    clearOnUpdate: false
  };

  state = {
    description: this.props.description,
    invalid: false
  };

  updateDescription = event => {
    const value = event.target.value;
    this.setState({
      description: value,
      invalid: value.length === 0
    });
  };

  handleSubmit = event => {
    event.preventDefault();
    const { description } = this.state;
    if (description.length > 0) {
      this.props.onUpdate(description);
      if (this.props.clearOnUpdate) {
        this.setState({ description: '' });
      }
    } else {
      this.setState({ invalid: true })
    }
  };

  render() {
    const { description, invalid } = this.state;
    const { className, invalidClassName } = this.props;
    const invalidStyle = invalid ? invalidClassName : '';
    return (
      <form onSubmit={this.handleSubmit}>
        <input
          className={`${className} ${invalidStyle}`}
          value={description}
          onChange={this.updateDescription}
        />
      </form>
    );
  }
}
