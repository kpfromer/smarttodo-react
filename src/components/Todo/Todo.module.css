.todo {
  position: relative;
  font-size: 24px;
  border-bottom: 1px solid #ededed;
}

.todo:last-child {
  border-bottom: none;
}

.todo.editing {
  border-bottom: none;
  padding: 0;
}

.todo.editing .edit {
  display: block;
  width: 506px;
  padding: 12px 16px;
  margin: 0 0 0 43px;
}

.todo.editing .view {
  display: none;
}

.todo .toggle {
  text-align: center;
  width: 40px;
  /* auto, since non-WebKit browsers doesn't support input styling */
  height: auto;
  position: absolute;
  top: 0;
  bottom: 0;
  margin: auto 0;
  border: none; /* Mobile Safari */
}

.todo .toggle {
  opacity: 0;
}

.todo .toggle + label {
  /*
      Firefox requires `#` to be escaped - https://bugzilla.mozilla.org/show_bug.cgi?id=922433
      IE and Edge requires *everything* to be escaped to render, so we do that instead of just the `#` - https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/7157459/
  */
  background: url('data:image/svg+xml;utf8,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2240%22%20height%3D%2240%22%20viewBox%3D%22-10%20-18%20100%20135%22%3E%3Ccircle%20cx%3D%2250%22%20cy%3D%2250%22%20r%3D%2250%22%20fill%3D%22none%22%20stroke%3D%22%23ededed%22%20stroke-width%3D%223%22/%3E%3C/svg%3E')
  no-repeat
  center left;
}

.todo .toggle:checked + label {
  background-image: url('data:image/svg+xml;utf8,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2240%22%20height%3D%2240%22%20viewBox%3D%22-10%20-18%20100%20135%22%3E%3Ccircle%20cx%3D%2250%22%20cy%3D%2250%22%20r%3D%2250%22%20fill%3D%22none%22%20stroke%3D%22%23bddad5%22%20stroke-width%3D%223%22/%3E%3Cpath%20fill%3D%22%235dc2af%22%20d%3D%22M72%2025L42%2071%2027%2056l-4%204%2020%2020%2034-52z%22/%3E%3C/svg%3E');
}

.todo label {
  word-break: break-all;
  padding: 15px 15px 15px 60px;
  display: block;
  line-height: 1.2;
  transition: color 0.4s;
}

.todo.completed label {
  color: #d9d9d9;
  text-decoration: line-through;
}

/* TODO: fix padding margin top */
.todo .destroy {
  display: none;
  position: absolute;
  top: 0;
  right: 10px;
  bottom: 0;
  width: 40px;
  height: 40px;
  font-size: 30px;
  color: #cc9a9a;
  margin: auto 0 11px;
  transition: color 0.2s ease-out;
}

.todo .destroy:hover {
  color: #af5b5e;
}

.todo .destroy:after {
  content: 'Ã—';
}

.todo:hover .destroy {
  display: block;
}

.todo .edit {
  display: none;
  transition: border-bottom 250ms;
}

.edit.invalid {
  border-bottom: 1px solid #ff3228;
}

.todo.editing:last-child {
  margin-bottom: -1px;
}