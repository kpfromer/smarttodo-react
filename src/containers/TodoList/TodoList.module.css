.list {
  background: #fff;
  padding: 1em 1em 0;
  display: flex;
  flex-direction: column;
}

.list li {
  flex-grow: 1;
}

.newTodo {
  position: relative;
  margin: 0;
  width: 100%;
  font-size: 24px;
  line-height: 1.4em;
  color: inherit;
  box-sizing: border-box;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;

  padding: 16px 16px 16px 60px;
  border: none;
  transition: border-bottom 250ms;
}

.newTodo.invalid {
  border-bottom: 1px solid #ff3228;
}