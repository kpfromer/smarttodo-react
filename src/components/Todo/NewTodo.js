import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import DescriptionForm from "./DescriptionForm";

export default class NewTodo extends PureComponent {

  static propTypes = {
    onCreate: PropTypes.func.isRequired
  };

  render() {
    return (
      <ListItem>
        <ListItemText>
          <DescriptionForm clearOnUpdate onUpdate={this.props.onCreate}/>
        </ListItemText>
      </ListItem>
    );
  }
}
