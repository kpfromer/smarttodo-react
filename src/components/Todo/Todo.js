import React, { Component } from "react";
import PropTypes from 'prop-types';

import styles from './Todo.module.scss'

import DescriptionForm from "./DescriptionForm";

export default class Todo extends Component {

  static propTypes = {
    description: PropTypes.string.isRequired,
    completed: PropTypes.bool.isRequired,
    onCheck: PropTypes.func.isRequired,
    onRemove: PropTypes.func.isRequired,
    onUpdate: PropTypes.func.isRequired
  };

  state = {
    editing: false
  };

  checkTodo = event => this.props.onCheck(event.target.checked);
  removeTodo = () => this.props.onRemove();

  startEditing = () =>
    this.setState({ editing: true });

  handleUpdate = description => {
    this.setState({ editing: false });
    this.props.onUpdate(description);
  };

  render() {
    const { description, completed } = this.props;

    const completedStyle = completed ? styles.completed : '';
    const editingStyle = this.state.editing ? styles.editing : '';

    return (
      <li className={`${styles.todo} ${completedStyle} ${editingStyle}`}>
        <div className={styles.view}>
          <input
            className={styles.toggle}
            type="checkbox"
            checked={completed}
            onChange={this.checkTodo}
          />
          <label onDoubleClick={this.startEditing}>{description}</label>
          <button className={styles.destroy} onClick={this.removeTodo}/>
        </div>
        <DescriptionForm
          className={styles.edit}
          invalidClassName={styles.invalid}
          description={description}
          onUpdate={this.handleUpdate}
        />
      </li>
    );
  }
}