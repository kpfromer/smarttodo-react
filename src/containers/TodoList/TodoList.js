import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import * as TodoActionCreators from '../../actions/todo';

import styles from './TodoList.module.scss';

import Todo from "../../components/Todo/Todo";
import DescriptionForm from "../../components/Todo/DescriptionForm";

export class TodoList extends Component {

  static propTypes = {
    todos: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      completed: PropTypes.bool.isRequired
    })),
    addTodo: PropTypes.func.isRequired,
    updateTodo: PropTypes.func.isRequired,
    removeTodo: PropTypes.func.isRequired
  };

  updateTodoByProp = propName => todo => value =>
    this.props.updateTodo(todo, {
      ...todo,
      [propName]: value
    }, todo.id);

  checkTodo = this.updateTodoByProp('completed');
  updateDescription = this.updateTodoByProp('description');
  removeTodo = (todo, index) => () => this.props.removeTodo(todo, index, todo.id);

  render() {
    const { todos, addTodo } = this.props;

    const todoItems = todos.map((todo, index) =>
      <Todo
        key={todo.id}
        description={todo.description}
        completed={todo.completed}
        onCheck={this.checkTodo(todo)}
        onUpdate={this.updateDescription(todo)}
        onRemove={this.removeTodo(todo, index)}
      />
    );
    return (
      <ul className={styles.list}>
        {todoItems}
        <DescriptionForm
          clearOnUpdate
          className={styles.newTodo}
          invalidClassName={styles.invalid}
          onUpdate={addTodo}
        />
      </ul>
    );
  }
}

export const mapStateToProps = state => ({
  todos: state.todos.todos
});

const staticActions = {
  addTodo: TodoActionCreators.addTodo,
  updateTodo: TodoActionCreators.updateTodo,
  removeTodo: TodoActionCreators.removeTodo
};

export default connect(mapStateToProps, staticActions)(TodoList);