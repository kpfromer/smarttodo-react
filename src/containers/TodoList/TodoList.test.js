import { mapStateToProps, TodoList } from "./TodoList";
import { SetupComponent } from "react-component-setup";
import List from '@material-ui/core/List';
import CircularProgress from '@material-ui/core/CircularProgress';
import Todo from "../../components/Todo/Todo";
import NewTodo from "../../components/Todo/NewTodo";

const todos = [
  {
    id: '10',
    description: 'math quiz',
    completed: false
  },
  {
    id: '11',
    description: 'computer science test',
    completed: true
  }
];

const { shallow: setup } = SetupComponent({
  component: TodoList,
  defaultProps: {
    addTodo: () => {},
    updateTodo: () => {},
    removeTodo: () => {},
    loadTodos: () => {},
    todos,
    isLoading: false
  }
});

describe('TodoList', () => {
  let wrapper;

  it('loads todos on creation', () => {
    const mockLoadTodos = jest.fn();
    setup({
      loadTodos: mockLoadTodos
    });

    expect(mockLoadTodos).toHaveBeenCalled();
  });

  it('renders a spinner while fetching todos', () => {
    ({ wrapper } = setup({
      isLoading: true
    }));
    expect(wrapper.find(CircularProgress)).toExist();
    expect(wrapper.find(List)).not.toExist();
  });

  describe('given `todos`', () => {
    beforeEach(() => {
      ({ wrapper } = setup());
    });

    it('renders a list', () => {
      expect(wrapper.is(List)).toBe(true);
    });

    it('renders a list of Todos', () => {
      expect(wrapper.find(List).find(Todo)).toMatchSnapshot();
    });
  });

  it('renders a NewTodo in the list', () => {
    ({ wrapper } = setup());

    expect(wrapper.find(NewTodo)).toExist();
  });

  describe('on NewTodo create', () => {
    let mockAddTodo;

    beforeEach(() => {
      mockAddTodo = jest.fn();

      ({ wrapper } = setup({
        addTodo: mockAddTodo
      }));

      wrapper.find(NewTodo).props().onCreate('new description');
      wrapper.update();
    });

    it('creates a new Todo', () => {
      expect(mockAddTodo).toHaveBeenCalledWith('new description');
    });
  });

  describe('Todo', () => {
    let mockUpdateTodo, mockRemoveTodo;
    beforeEach(() => {
      mockUpdateTodo = jest.fn();
      mockRemoveTodo = jest.fn();

      ({ wrapper } = setup({
        updateTodo: mockUpdateTodo,
        removeTodo: mockRemoveTodo
      }));
    });
    describe('on check', () => {
      it('updates the todo\'s completed value', () => {
        wrapper.find(Todo).first().props().onCheck(true);
        expect(mockUpdateTodo).toHaveBeenCalledWith(
          {
            id: '10',
            description: 'math quiz',
            completed: false
          },
          {
            id: '10',
            description: 'math quiz',
            completed: true
          },
          '10'
        );
      });
    });

    describe('on remove', () => {
      it('removes the todo', () => {
        const position = 1;
        wrapper.find(Todo).at(position).props().onRemove();
        expect(mockRemoveTodo).toHaveBeenCalledWith(todos[position], position, todos[position].id);
      });
    });

    describe('on update', () => {
      it('updates the todo\'s description', () => {
        const description = 'new description';
        wrapper.find(Todo).first().props().onUpdate(description);
        expect(mockUpdateTodo).toHaveBeenCalledWith(
          {
            id: '10',
            description: 'math quiz',
            completed: false
          },
          {
            id: '10',
            description: description,
            completed: false
          },
          '10'
        );
      });
    });
  });
});

describe('mapStateToProps', () => {
  let valid;

  beforeEach(() => {
    valid = {
      todos: [],
      isFetching: false
    }
  });

  it('returns todos', () => {
    const mockTodos = [
      {
        id: '1',
        description: 'hello',
        completed: false
      },
      {
        id: '2',
        description: 'world',
        completed: true
      }
    ];
    const mapped = mapStateToProps({
      todos: {
        ...valid,
        todos: mockTodos
      }
    });

    expect(mapped).toMatchObject({
      todos: mockTodos
    });
  });
  it('returns isLoading', () => {
    const isFetching = true;
    const mapped = mapStateToProps({
      todos: {
        ...valid,
        isFetching
      }
    });
    expect(mapped).toMatchObject({
      isLoading: isFetching
    })
  });
});
