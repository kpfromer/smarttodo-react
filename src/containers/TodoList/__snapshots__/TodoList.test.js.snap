// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList given \`todos\` renders a list of Todos 1`] = `
Array [
  <Todo
    completed={false}
    description="math quiz"
    key="10"
    onCheck={[Function]}
    onRemove={[Function]}
    onUpdate={[Function]}
  />,
  <Todo
    completed={true}
    description="computer science test"
    key="11"
    onCheck={[Function]}
    onRemove={[Function]}
    onUpdate={[Function]}
  />,
]
`;

exports[`TodoList renders a DescriptionForm in the list 1`] = `
<DescriptionForm
  className="newTodo"
  clearOnUpdate={true}
  description=""
  invalidClassName="invalid"
  onUpdate={[Function]}
/>
`;
